// @emmett.deen/linear-mcp-server/dist/controllers/context.controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContext = void 0;
const linear_service_1 = require("../services/linear.service");
/**
 * Get context from Linear API
 * @route GET /api/context
 */
const getContext = async (req, res) => {
    try {
        // Get token from middleware
        const token = req.headers.authorization?.replace('Bearer ', '') || '';
        // Get query parameters
        const { teamIds, projectIds, assigneeIds, states, priorities, searchQuery, dateRange, first, after, } = req.query;
        // Initialize Linear service
        const linearService = new linear_service_1.LinearService(token);
        // Parse query parameters
        const filters = {
            teamIds: teamIds ? teamIds.split(',') : undefined,
            projectIds: projectIds ? projectIds.split(',') : undefined,
            assigneeIds: assigneeIds ? assigneeIds.split(',') : undefined,
            states: states ? states.split(',') : undefined,
            priorities: priorities ? priorities.split(',').map(Number) : undefined,
            searchQuery,
            dateRange: dateRange ? JSON.parse(dateRange) : undefined,
        };
        const pagination = {
            first: first ? Number(first) : 25,
            after,
        };
        // Get user info
        const userInfo = await linearService.getUserInfo();
        // Get issues
        const { issues, totalCount, pageInfo } = await linearService.getIssues(filters, pagination);
        // Get projects (filtering by teamIds if provided)
        const projects = await linearService.getProjects(filters.teamIds);
        // Get teams
        const teams = await linearService.getTeams();
        // Get users
        const users = await linearService.getUsers();
        // Format response
        const response = {
            context: {
                issues,
                projects,
                teams,
                users,
                userInfo,
            },
            metadata: {
                totalIssues: totalCount,
                hasNextPage: pageInfo.hasNextPage,
                endCursor: pageInfo.endCursor,
            },
        };
        res.status(200).json(response);
    }
    catch (error) {
        console.error('Error getting context:', error);
        res.status(500).json({
            error: 'Failed to get context',
            message: error.message,
        });
    }
};
exports.getContext = getContext;
