// @emmett.deen/linear-mcp-server/dist/services/linear.service.js
import { ActionType, } from '../types/mcp.js';
export class LinearService {
    // Use definite assignment assertion to tell TypeScript this will be initialized
    client;
    apiToken;
    clientInitialized = false;
    constructor(apiToken) {
        this.apiToken = apiToken;
        // Don't call async method from constructor
        // We'll ensure initClient is called before any operations
    }
    /**
     * Initialize the GraphQL client
     */
    async initClient(apiToken = this.apiToken) {
        try {
            // Dynamic import of the GraphQLClient
            const { GraphQLClient } = await import('graphql-request');
            this.client = new GraphQLClient('https://api.linear.app/graphql', {
                headers: {
                    authorization: apiToken,
                },
            });
            this.clientInitialized = true;
        }
        catch (error) {
            console.error('Failed to initialize GraphQL client:', error);
            throw error;
        }
    }
    /**
     * Ensure client is initialized before any operation
     */
    async ensureClient() {
        if (!this.clientInitialized) {
            await this.initClient();
        }
    }
    /**
     * Get current user info
     */
    async getUserInfo() {
        // Make sure client is initialized
        await this.ensureClient();
        const query = `
      query GetUserInfo {
        viewer {
          id
          name
          email
          organization {
            id
            name
          }
          teams {
            nodes {
              id
              name
              description
              key
            }
          }
        }
      }
    `;
        const data = await this.client.request(query);
        return {
            id: data.viewer.id,
            name: data.viewer.name,
            email: data.viewer.email,
            organization: {
                id: data.viewer.organization.id,
                name: data.viewer.organization.name,
            },
            teams: data.viewer.teams.nodes.map((team) => ({
                id: team.id,
                name: team.name,
                description: team.description,
                key: team.key,
            })),
        };
    }
    /**
     * Get issues with specified filters
     */
    async getIssues(filters = {}, pagination = { first: 25 }) {
        await this.ensureClient();
        const { teamIds, projectIds, assigneeIds, states, priorities, searchQuery, dateRange } = filters;
        const { first, after } = pagination;
        let filterConditions = '';
        if (teamIds?.length) {
            filterConditions += `teamId: { in: [${teamIds.map((id) => `"${id}"`).join(', ')}] }`;
        }
        if (projectIds?.length) {
            if (filterConditions)
                filterConditions += ', ';
            filterConditions += `projectId: { in: [${projectIds.map((id) => `"${id}"`).join(', ')}] }`;
        }
        if (assigneeIds?.length) {
            if (filterConditions)
                filterConditions += ', ';
            filterConditions += `assigneeId: { in: [${assigneeIds.map((id) => `"${id}"`).join(', ')}] }`;
        }
        if (states?.length) {
            if (filterConditions)
                filterConditions += ', ';
            filterConditions += `stateId: { in: [${states.map((id) => `"${id}"`).join(', ')}] }`;
        }
        if (priorities?.length) {
            if (filterConditions)
                filterConditions += ', ';
            filterConditions += `priority: { in: [${priorities.join(', ')}] }`;
        }
        if (searchQuery) {
            if (filterConditions)
                filterConditions += ', ';
            filterConditions += `title: { contains: "${searchQuery}" }`;
        }
        if (dateRange?.start && dateRange?.end) {
            if (filterConditions)
                filterConditions += ', ';
            filterConditions += `createdAt: { gte: "${dateRange.start}", lte: "${dateRange.end}" }`;
        }
        // Build the filter object for the query
        const filterString = filterConditions ? `(filter: { ${filterConditions} })` : '';
        const paginationString = `first: ${first}${after ? `, after: "${after}"` : ''}`;
        const query = `
      query GetIssues {
        issues${filterString} {
          totalCount
          pageInfo {
            hasNextPage
            endCursor
          }
          nodes {
            id
            title
            description
            priority
            createdAt
            updatedAt
            url
            state {
              id
              name
              type
            }
            assignee {
              id
              name
            }
            team {
              id
              name
            }
            project {
              id
              name
            }
            comments(first: 5) {
              nodes {
                id
                body
                createdAt
                user {
                  id
                  name
                }
              }
            }
          }
        }
      }
    `;
        const data = await this.client.request(query);
        return {
            issues: data.issues.nodes.map((issue) => ({
                id: issue.id,
                title: issue.title,
                description: issue.description,
                priority: issue.priority,
                createdAt: issue.createdAt,
                updatedAt: issue.updatedAt,
                url: issue.url,
                state: issue.state && {
                    id: issue.state.id,
                    name: issue.state.name,
                    type: issue.state.type,
                },
                assignee: issue.assignee && {
                    id: issue.assignee.id,
                    name: issue.assignee.name,
                },
                team: issue.team && {
                    id: issue.team.id,
                    name: issue.team.name,
                },
                project: issue.project && {
                    id: issue.project.id,
                    name: issue.project.name,
                },
                comments: issue.comments?.nodes.map((comment) => ({
                    id: comment.id,
                    body: comment.body,
                    createdAt: comment.createdAt,
                    user: {
                        id: comment.user.id,
                        name: comment.user.name,
                    },
                })),
            })),
            totalCount: data.issues.totalCount,
            pageInfo: data.issues.pageInfo,
        };
    }
    /**
     * Get projects
     */
    async getProjects(teamIds = []) {
        await this.ensureClient();
        let filterCondition = '';
        if (teamIds.length > 0) {
            filterCondition = `(filter: { teamId: { in: [${teamIds.map(id => `"${id}"`).join(', ')}] } })`;
        }
        const query = `
      query GetProjects {
        projects${filterCondition} {
          nodes {
            id
            name
            description
            state
            startDate
            targetDate
            team {
              id
              name
            }
            issueCountHistory {
              total
              completed
            }
            url
          }
        }
      }
    `;
        const data = await this.client.request(query);
        return data.projects.nodes.map((project) => ({
            id: project.id,
            name: project.name,
            description: project.description,
            state: project.state,
            startDate: project.startDate,
            targetDate: project.targetDate,
            team: project.team && {
                id: project.team.id,
                name: project.team.name,
            },
            issueCount: project.issueCountHistory?.total || 0,
            completedIssueCount: project.issueCountHistory?.completed || 0,
            url: project.url,
        }));
    }
    /**
     * Get teams
     */
    async getTeams() {
        await this.ensureClient();
        const query = `
      query GetTeams {
        teams {
          nodes {
            id
            name
            description
            key
            members {
              nodes {
                id
                name
                displayName
                email
                avatarUrl
              }
            }
          }
        }
      }
    `;
        const data = await this.client.request(query);
        return data.teams.nodes.map((team) => ({
            id: team.id,
            name: team.name,
            description: team.description,
            key: team.key,
            members: team.members.nodes.map((member) => ({
                id: member.id,
                name: member.name,
                displayName: member.displayName,
                email: member.email,
                avatarUrl: member.avatarUrl,
            })),
        }));
    }
    /**
     * Get users
     */
    async getUsers() {
        await this.ensureClient();
        const query = `
      query GetUsers {
        users {
          nodes {
            id
            name
            displayName
            email
            avatarUrl
          }
        }
      }
    `;
        const data = await this.client.request(query);
        return data.users.nodes.map((user) => ({
            id: user.id,
            name: user.name,
            displayName: user.displayName,
            email: user.email,
            avatarUrl: user.avatarUrl,
        }));
    }
    /**
     * Perform action in Linear
     */
    async performAction(type, payload) {
        await this.ensureClient();
        switch (type) {
            case ActionType.CREATE_ISSUE:
                return this.createIssue(payload);
            case ActionType.UPDATE_ISSUE:
                return this.updateIssue(payload);
            case ActionType.CHANGE_ISSUE_STATUS:
                return this.changeIssueStatus(payload);
            case ActionType.ASSIGN_ISSUE:
                return this.assignIssue(payload);
            case ActionType.ADD_COMMENT:
                return this.addComment(payload);
            case ActionType.CREATE_PROJECT:
                return this.createProject(payload);
            case ActionType.CREATE_TEAM:
                return this.createTeam(payload);
            default:
                throw new Error(`Unsupported action type: ${type}`);
        }
    }
    /**
     * Create a new issue
     */
    async createIssue(payload) {
        await this.ensureClient();
        const { title, description, teamId, projectId, assigneeId, stateId, priority } = payload;
        const mutation = `
      mutation CreateIssue($title: String!, $description: String, $teamId: String!, $projectId: String, $assigneeId: String, $stateId: String, $priority: Int) {
        issueCreate(input: {
          title: $title,
          description: $description,
          teamId: $teamId,
          projectId: $projectId,
          assigneeId: $assigneeId,
          stateId: $stateId,
          priority: $priority
        }) {
          success
          issue {
            id
            title
            url
          }
        }
      }
    `;
        const variables = {
            title,
            description,
            teamId,
            projectId,
            assigneeId,
            stateId,
            priority,
        };
        return this.client.request(mutation, variables);
    }
    /**
     * Update an existing issue
     */
    async updateIssue(payload) {
        await this.ensureClient();
        const { id, title, description, projectId, priority } = payload;
        const mutation = `
      mutation UpdateIssue($id: String!, $title: String, $description: String, $projectId: String, $priority: Int) {
        issueUpdate(id: $id, input: {
          title: $title,
          description: $description,
          projectId: $projectId,
          priority: $priority
        }) {
          success
          issue {
            id
            title
            url
          }
        }
      }
    `;
        const variables = {
            id,
            title,
            description,
            projectId,
            priority,
        };
        return this.client.request(mutation, variables);
    }
    /**
     * Change issue status
     */
    async changeIssueStatus(payload) {
        await this.ensureClient();
        const { issueId, stateId } = payload;
        const mutation = `
      mutation ChangeIssueStatus($issueId: String!, $stateId: String!) {
        issueUpdate(id: $issueId, input: {
          stateId: $stateId
        }) {
          success
          issue {
            id
            title
            state {
              id
              name
            }
          }
        }
      }
    `;
        const variables = {
            issueId,
            stateId,
        };
        return this.client.request(mutation, variables);
    }
    /**
     * Assign issue to user
     */
    async assignIssue(payload) {
        await this.ensureClient();
        const { issueId, assigneeId } = payload;
        const mutation = `
      mutation AssignIssue($issueId: String!, $assigneeId: String) {
        issueUpdate(id: $issueId, input: {
          assigneeId: $assigneeId
        }) {
          success
          issue {
            id
            title
            assignee {
              id
              name
            }
          }
        }
      }
    `;
        const variables = {
            issueId,
            assigneeId,
        };
        return this.client.request(mutation, variables);
    }
    /**
     * Add comment to issue
     */
    async addComment(payload) {
        await this.ensureClient();
        const { issueId, body } = payload;
        const mutation = `
      mutation AddComment($issueId: String!, $body: String!) {
        commentCreate(input: {
          issueId: $issueId,
          body: $body
        }) {
          success
          comment {
            id
            body
          }
        }
      }
    `;
        const variables = {
            issueId,
            body,
        };
        return this.client.request(mutation, variables);
    }
    /**
     * Create a new project
     */
    async createProject(payload) {
        await this.ensureClient();
        const { name, description, teamId, state, startDate, targetDate } = payload;
        const mutation = `
      mutation CreateProject($name: String!, $description: String, $teamId: String!, $state: String, $startDate: TimelessDate, $targetDate: TimelessDate) {
        projectCreate(input: {
          name: $name,
          description: $description,
          teamId: $teamId,
          state: $state,
          startDate: $startDate,
          targetDate: $targetDate
        }) {
          success
          project {
            id
            name
            url
          }
        }
      }
    `;
        const variables = {
            name,
            description,
            teamId,
            state,
            startDate,
            targetDate,
        };
        return this.client.request(mutation, variables);
    }
    /**
     * Create a new team
     */
    async createTeam(payload) {
        await this.ensureClient();
        const { name, description, key } = payload;
        const mutation = `
      mutation CreateTeam($name: String!, $description: String, $key: String!) {
        teamCreate(input: {
          name: $name,
          description: $description,
          key: $key
        }) {
          success
          team {
            id
            name
            key
          }
        }
      }
    `;
        const variables = {
            name,
            description,
            key,
        };
        return this.client.request(mutation, variables);
    }
}
