// @emmett.deen/linear-mcp-server/dist/tools/type-guards.js
/**
 * Type guard for linear_getIssues tool arguments
 */
export function isGetIssuesArgs(args) {
    return (typeof args === "object" &&
        args !== null &&
        (!("limit" in args) || typeof args.limit === "number"));
}
/**
 * Type guard for linear_getIssueById tool arguments
 */
export function isGetIssueByIdArgs(args) {
    return (typeof args === "object" &&
        args !== null &&
        "id" in args &&
        typeof args.id === "string");
}
/**
 * Type guard for linear_searchIssues tool arguments
 */
export function isSearchIssuesArgs(args) {
    return (typeof args === "object" &&
        args !== null &&
        (!("query" in args) || typeof args.query === "string") &&
        (!("teamId" in args) || typeof args.teamId === "string") &&
        (!("assigneeId" in args) || typeof args.assigneeId === "string") &&
        (!("projectId" in args) || typeof args.projectId === "string") &&
        (!("states" in args) || Array.isArray(args.states)) &&
        (!("limit" in args) || typeof args.limit === "number"));
}
/**
 * Type guard for linear_createIssue tool arguments
 */
export function isCreateIssueArgs(args) {
    return (typeof args === "object" &&
        args !== null &&
        "title" in args &&
        typeof args.title === "string" &&
        "teamId" in args &&
        typeof args.teamId === "string");
}
/**
 * Type guard for linear_updateIssue tool arguments
 */
export function isUpdateIssueArgs(args) {
    return (typeof args === "object" &&
        args !== null &&
        "id" in args &&
        typeof args.id === "string");
}
/**
 * Type guard for linear_createComment tool arguments
 */
export function isCreateCommentArgs(args) {
    return (typeof args === "object" &&
        args !== null &&
        "issueId" in args &&
        typeof args.issueId === "string" &&
        "body" in args &&
        typeof args.body === "string");
}
/**
 * Type guard for linear_createProject tool arguments
 */
export function isCreateProjectArgs(args) {
    return (typeof args === "object" &&
        args !== null &&
        "name" in args &&
        typeof args.name === "string" &&
        "teamIds" in args &&
        Array.isArray(args.teamIds));
}
/**
 * Type guard for linear_addIssueLabel tool arguments
 */
export function isAddIssueLabelArgs(args) {
    return (typeof args === "object" &&
        args !== null &&
        "issueId" in args &&
        typeof args.issueId === "string" &&
        "labelId" in args &&
        typeof args.labelId === "string");
}
/**
 * Type guard for linear_removeIssueLabel tool arguments
 */
export function isRemoveIssueLabelArgs(args) {
    return (typeof args === "object" &&
        args !== null &&
        "issueId" in args &&
        typeof args.issueId === "string" &&
        "labelId" in args &&
        typeof args.labelId === "string");
}
