// @emmett.deen/linear-mcp-server/dist/utils/error.util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatError = exports.asyncHandler = exports.APIError = void 0;
/**
 * Custom API error class
 */
class APIError extends Error {
    statusCode;
    constructor(message, statusCode = 500) {
        super(message);
        this.name = this.constructor.name;
        this.statusCode = statusCode;
        Error.captureStackTrace(this, this.constructor);
    }
}
exports.APIError = APIError;
/**
 * Handler for async route functions
 */
const asyncHandler = (fn) => {
    return async (req, res, next) => {
        try {
            await fn(req, res, next);
        }
        catch (error) {
            next(error);
        }
    };
};
exports.asyncHandler = asyncHandler;
/**
 * Format error response
 */
const formatError = (error) => {
    return {
        error: error.name || 'Error',
        message: error.message || 'An unknown error occurred',
        status: error.statusCode || 500,
        ...(process.env.NODE_ENV === 'development' && { stack: error.stack }),
    };
};
exports.formatError = formatError;
